ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+      reserveFile = createReserveFile(sizeInBytes);
+    }
+    if (destinationFile.exists() || !reserveFile.renameTo(destinationFile)) {
+      reserveFiles.add(sizeInBytes, reserveFile);
+      throw new IOException("Error while moving reserve file: " + reserveFile.getAbsolutePath() + " to location: "
+          + destinationFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Return a file to the pool. The user must keep track of the size of the file allocated.
+   * @param fileToReturn the file to return to the pool.
+   * @param sizeInBytes the size of the file to return.
+   * @throws IOException if the file to return does not exist or cannot be cleaned or recreated correctly.
+   */
+  void free(File fileToReturn, long sizeInBytes) throws IOException {
+    if (!poolInitialized) {
+      LOGGER.debug("Freeing segment before pool is fully initialized");
+    }
+    // For now, we delete the file and create a new one. Newer linux kernel versions support
+    // additional fallocate flags, which will be useful for cleaning up returned files.
+    if (!fileToReturn.delete()) {
+      throw new IOException("Error while cleaning up returned file: " + fileToReturn.getAbsolutePath());
+    }
+    fileToReturn = createReserveFile(sizeInBytes);
+    reserveFiles.add(sizeInBytes, fileToReturn);
+  }
+
+  /**
+   * @return the number of swap segments that should be created in the pool for each segment size.
+   */
+  long getSwapSegmentsPerSize() {
+    return requiredSwapSegmentsPerSize;
+  }
+
+  /**
+   * Inventory existing reserve directories and generate a {@link ReserveFileMap}.
+   * @return a populated {@link ReserveFileMap}
+   */
+  private ReserveFileMap inventoryExistingReserveFiles() {
+    ReserveFileMap reserveFiles = new ReserveFileMap();
+    File[] fileSizeDirs = reserveDir.listFiles(FILE_SIZE_DIR_FILTER);
+    if (fileSizeDirs != null) {
+      for (File fileSizeDir : fileSizeDirs) {
+        long sizeInBytes = Long.parseLong(fileSizeDir.getName());
+        File[] reserveFilesForSize = fileSizeDir.listFiles(RESERVE_FILE_FILTER);
+        if (reserveFilesForSize != null) {
+          for (File reserveFile : reserveFilesForSize) {
+            reserveFiles.add(sizeInBytes, reserveFile);
+          }
+        }
+      }
+    }
+    return reserveFiles;
+  }
+
+  /**
+   * Delete the currently-present reserve files that are not required by {@code overallRequirements}.
+   * @param overallRequirements a map between segment sizes in bytes and the number of segments needed for that size.
+   * @throws IOException
+   */
+  private void deleteUnneededSegments(Map<Long, Long> overallRequirements) throws IOException {
+    for (long sizeInBytes : reserveFiles.getFileSizeSet()) {
+      Long segmentsNeeded = overallRequirements.get(sizeInBytes);
+      if (segmentsNeeded == null || segmentsNeeded == 0) {
+        File dirToDelete = new File(reserveDir, Long.toString(sizeInBytes));
+        if (!Utils.deleteDirectory(dirToDelete)) {
```

Comment:
if dirToDelete does not exist, Utils.deleteDirectory will throw an exception I guess